workflow/controller/dag_test.go

```diff
@@ -1764,6 +1760,12 @@ func TestOnExitNonLeaf(t *testing.T) {
 	if assert.NotNil(t, retryNode) {
 		assert.Equal(t, wfv1.NodePending, retryNode.Phase)
 	}
+
+	assert.Nil(t, woc.getNodeByName("exit-handler-bug-example.step-3"))
```

Comment:
This modified test tests for the new expected behavior and in particular this line: `step-3` should not run until the exit handler has finished